File: MoonSimulation/Renderers/SkySceneRenderer.cs
using SkiaSharp;

namespace MoonSimulation.Renderers;

/// <summary>
/// Orchestrates the "view from Earth" sky scene — what the observer sees.
/// Instance-based because it shares the starfield state.
/// </summary>
public class SkySceneRenderer
{
    private readonly StarfieldRenderer _starfield;

    public SkySceneRenderer(StarfieldRenderer starfield)
    {
        _starfield = starfield;
    }

    /// <summary>
    /// Draws the complete Earth sky view scene.
    /// </summary>
    public void DrawEarthSkyView(SKCanvas canvas, SKRect bounds,
        double sunElevation, double moonAngleDeg, double earthRotationDeg, double illumination,
        (string Name, string Emoji) phase, (string Name, string Emoji) timeOfDay, double hourOfDay)
    {
        float w = bounds.Width;
        float h = bounds.Height;
        float groundY = bounds.Top + h * 0.75f;

        // Sky gradient based on sun elevation
        DrawSkyGradient(canvas, bounds, groundY, sunElevation);

        // Ground
        DrawGround(canvas, bounds, groundY, sunElevation);

        // Sun in the sky (when above horizon)
        if (sunElevation > -0.1)
        {
            float sunArc = (float)(1.0 - (earthRotationDeg % 360) / 180.0);
            float sunSkyX = bounds.Left + w * 0.5f + w * 0.35f * (float)Math.Cos(earthRotationDeg * Math.PI / 180.0);
            float sunSkyY = groundY - (groundY - bounds.Top) * 0.8f * Math.Max(0, (float)sunElevation);

            if (sunElevation > 0)
            {
                float sunR = Math.Min(w, h) * 0.06f;
                // Sun glow
                using var glowPaint = new SKPaint
                {
                    IsAntialias = true,
                    Color = new SKColor(255, 230, 100, (byte)(40 * sunElevation)),
                    MaskFilter = SKMaskFilter.CreateBlur(SKBlurStyle.Normal, sunR * 2)
                };
                canvas.DrawCircle(sunSkyX, sunSkyY, sunR * 3, glowPaint);

                // Sun disk
                using var sunPaint = new SKPaint { IsAntialias = true };
                using var sunShader = SKShader.CreateRadialGradient(
                    new SKPoint(sunSkyX, sunSkyY), sunR,
                    new SKColor[] { new(255, 255, 230), new(255, 200, 50) },
                    null, SKShaderTileMode.Clamp);
                sunPaint.Shader = sunShader;
                canvas.DrawCircle(sunSkyX, sunSkyY, sunR, sunPaint);
            }
        }

        // Moon in the sky — visible when above observer's horizon AND sufficiently illuminated.
        // moonAngleDeg: Moon's orbital angle (0° = between Sun and Earth, 180° = far side)
        // earthRotationDeg: observer's facing (0° = facing Sun/noon, 180° = away/midnight)
        // The angular difference tells us how far the Moon is from the observer's zenith.
        double moonObserverAngle = (moonAngleDeg - earthRotationDeg) * Math.PI / 180.0;
        double moonElevation = Math.Cos(moonObserverAngle);

        // Moon is only visually noticeable if:
        // 1. Above the horizon (elevation > 0)
        // 2. Has enough illumination to be seen (near New Moon it's invisible)
        // 3. During daytime, needs significant illumination to stand out against bright sky
        double minIlluminationToSee = sunElevation > 0.2 ? 0.25 : 0.03;
        bool moonVisible = moonElevation > 0.05 && illumination > minIlluminationToSee;

        if (moonVisible)
        {
            float moonSkyX = bounds.Left + w * 0.5f - w * 0.3f * (float)Math.Sin(moonObserverAngle);
            float moonSkyY = groundY - (groundY - bounds.Top) * 0.7f * (float)moonElevation;
            float moonR = Math.Min(w, h) * 0.055f;

            // Fade the moon based on how close it is to the horizon and its illumination
            float moonAlpha = (float)Math.Min(1.0, moonElevation * 2) * (float)Math.Min(1.0, illumination * 3);

            DrawMiniMoonPhase(canvas, new SKPoint(moonSkyX, moonSkyY), moonR,
                moonAngleDeg, sunElevation, moonAlpha);
        }

        // Stars (only visible at night)
        if (sunElevation < 0.2)
        {
            float starAlpha = (float)Math.Clamp((0.2 - sunElevation) / 0.4, 0, 1);
            DrawSkyStars(canvas, bounds, groundY, starAlpha);

            // Shooting star at night
            if (sunElevation < -0.2)
                DrawShootingStar(canvas, bounds, groundY, _starfield.TwinkleFrame);
        }

        // Clouds during daytime
        SceneryRenderer.DrawClouds(canvas, bounds, groundY, sunElevation, _starfield.TwinkleFrame);

        // Little house on the ground
        SceneryRenderer.DrawLittleHouse(canvas, bounds, groundY, sunElevation);

        // Time label
        int hours = (int)hourOfDay;
        int mins = (int)((hourOfDay - hours) * 60);
        string timeStr = $"{hours:D2}:{mins:D2}";

        SphereRenderer.DrawLabel(canvas, timeStr,
            new SKPoint(bounds.Left + w * 0.5f, bounds.Top + 22),
            16, new SKColor(230, 230, 230));

        // Day/Night label
        SphereRenderer.DrawLabel(canvas, timeOfDay.Name,
            new SKPoint(bounds.Left + w * 0.5f, bounds.Top + 44),
            14, new SKColor(200, 200, 220));
    }

    /// <summary>
    /// Draws a small moon with the correct phase for the sky view.
    /// Delegates shadow to MoonPhaseRenderer.DrawPhaseOverlay.
    /// </summary>
    public static void DrawMiniMoonPhase(SKCanvas canvas, SKPoint center, float radius,
        double moonAngleDeg, double sunElevation, float alpha = 1f)
    {
        // Moon glow (brighter at night), scaled by alpha
        float glowAlpha = (sunElevation < 0 ? 0.4f : 0.15f) * alpha;
        using var glowPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(200, 200, 220, (byte)(glowAlpha * 255)),
            MaskFilter = SKMaskFilter.CreateBlur(SKBlurStyle.Normal, radius * 0.8f)
        };
        canvas.DrawCircle(center, radius * 1.5f, glowPaint);

        // Moon disk
        byte diskAlpha = (byte)(alpha * 255);
        using var moonPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(220, 220, 215, diskAlpha)
        };
        canvas.DrawCircle(center, radius, moonPaint);

        // Phase shadow — uses the shared terminator algorithm
        byte shadowAlpha = (byte)(alpha * (sunElevation < 0 ? 230 : 200));
        var shadowColor = sunElevation < 0
            ? new SKColor(5, 5, 25, shadowAlpha)
            : new SKColor(100, 140, 200, shadowAlpha);
        MoonPhaseRenderer.DrawPhaseOverlay(canvas, center, radius, moonAngleDeg, shadowColor);
    }

    public static void DrawSkyGradient(SKCanvas canvas, SKRect bounds, float groundY,
        double sunElevation)
    {
        SKColor topColor, bottomColor;

        if (sunElevation > 0.3)
        {
            topColor = new SKColor(40, 100, 220);
            bottomColor = new SKColor(135, 190, 250);
        }
        else if (sunElevation > 0.0)
        {
            float t = (float)(sunElevation / 0.3);
            topColor = SphereRenderer.BlendColor(new SKColor(30, 30, 80), new SKColor(40, 100, 220), t);
            bottomColor = SphereRenderer.BlendColor(new SKColor(220, 100, 40), new SKColor(135, 190, 250), t);
        }
        else if (sunElevation > -0.3)
        {
            float t = (float)((sunElevation + 0.3) / 0.3);
            topColor = SphereRenderer.BlendColor(new SKColor(8, 8, 35), new SKColor(30, 30, 80), t);
            bottomColor = SphereRenderer.BlendColor(new SKColor(20, 15, 40), new SKColor(220, 100, 40), t);
        }
        else
        {
            topColor = new SKColor(5, 5, 25);
            bottomColor = new SKColor(10, 10, 35);
        }

        using var skyPaint = new SKPaint();
        using var skyShader = SKShader.CreateLinearGradient(
            new SKPoint(bounds.MidX, bounds.Top),
            new SKPoint(bounds.MidX, groundY),
            new SKColor[] { topColor, bottomColor },
            null, SKShaderTileMode.Clamp);
        skyPaint.Shader = skyShader;
        canvas.DrawRect(new SKRect(bounds.Left, bounds.Top, bounds.Right, groundY), skyPaint);
    }

    public static void DrawGround(SKCanvas canvas, SKRect bounds, float groundY,
        double sunElevation)
    {
        SKColor groundTop, groundBottom;

        if (sunElevation > 0.1)
        {
            groundTop = new SKColor(40, 120, 40);
            groundBottom = new SKColor(30, 80, 30);
        }
        else if (sunElevation > -0.1)
        {
            float t = (float)((sunElevation + 0.1) / 0.2);
            groundTop = SphereRenderer.BlendColor(new SKColor(15, 40, 15), new SKColor(40, 120, 40), t);
            groundBottom = SphereRenderer.BlendColor(new SKColor(10, 25, 10), new SKColor(30, 80, 30), t);
        }
        else
        {
            groundTop = new SKColor(15, 40, 15);
            groundBottom = new SKColor(10, 25, 10);
        }

        using var groundPaint = new SKPaint();
        using var groundShader = SKShader.CreateLinearGradient(
            new SKPoint(bounds.MidX, groundY),
            new SKPoint(bounds.MidX, bounds.Bottom),
            new SKColor[] { groundTop, groundBottom },
            null, SKShaderTileMode.Clamp);
        groundPaint.Shader = groundShader;
        canvas.DrawRect(new SKRect(bounds.Left, groundY, bounds.Right, bounds.Bottom), groundPaint);

        // Horizon line glow during sunrise/sunset
        if (sunElevation > -0.1 && sunElevation < 0.2)
        {
            using var horizonPaint = new SKPaint
            {
                IsAntialias = true,
                Color = new SKColor(255, 180, 80, (byte)(60 * (1 - Math.Abs(sunElevation) * 5))),
                MaskFilter = SKMaskFilter.CreateBlur(SKBlurStyle.Normal, 8)
            };
            canvas.DrawLine(bounds.Left, groundY, bounds.Right, groundY, horizonPaint);
        }
    }

    public void DrawSkyStars(SKCanvas canvas, SKRect bounds, float groundY, float alpha)
    {
        if (_starfield.Stars == null) return;

        using var starPaint = new SKPaint { IsAntialias = true };

        for (int i = 0; i < 60; i++)
        {
            float x = _starfield.Stars[i].X % bounds.Width + bounds.Left;
            float y = _starfield.Stars[i].Y % (groundY - bounds.Top) + bounds.Top;

            float twinkle = _starfield.StarBrightness![i] *
                (0.7f + 0.3f * MathF.Sin((_starfield.TwinkleFrame + i * 37) * 0.02f));
            byte a = (byte)(twinkle * alpha * 255);

            starPaint.Color = new SKColor(255, 255, 240, a);
            canvas.DrawCircle(x, y, _starfield.StarBrightness[i] > 0.8f ? 1.5f : 0.8f, starPaint);
        }
    }

    /// <summary>
    /// Draws a shooting star animation at night.
    /// </summary>
    public static void DrawShootingStar(SKCanvas canvas, SKRect bounds, float groundY, long frame)
    {
        int cycle = (int)(frame % 300);
        if (cycle > 30) return;

        float progress = cycle / 30f;
        float startX = bounds.Left + bounds.Width * 0.7f;
        float startY = bounds.Top + 20;
        float endX = bounds.Left + bounds.Width * 0.3f;
        float endY = groundY * 0.4f;

        float curX = startX + (endX - startX) * progress;
        float curY = startY + (endY - startY) * progress;

        // Tail
        float tailLen = 30f * (1 - progress);
        float tailX = curX - (endX - startX) / (endY - startY) * -tailLen;
        float tailY = curY - tailLen;

        using var starPaint = new SKPaint
        {
            IsAntialias = true,
            Style = SKPaintStyle.Stroke,
            StrokeWidth = 2f,
            StrokeCap = SKStrokeCap.Round
        };
        using var shader = SKShader.CreateLinearGradient(
            new SKPoint(tailX, tailY), new SKPoint(curX, curY),
            new SKColor[] { SKColors.Transparent, new SKColor(255, 255, 220, (byte)(200 * (1 - progress))) },
            null, SKShaderTileMode.Clamp);
        starPaint.Shader = shader;
        canvas.DrawLine(tailX, tailY, curX, curY, starPaint);

        // Head glow
        using var headPaint = new SKPaint
        {
            IsAntialias = true,
            Color = new SKColor(255, 255, 230, (byte)(255 * (1 - progress))),
            MaskFilter = SKMaskFilter.CreateBlur(SKBlurStyle.Normal, 3)
        };
        canvas.DrawCircle(curX, curY, 2, headPaint);
    }
}

File: web/js/renderers/skySceneRenderer.js
import { blendColor } from './sphereRenderer.js';
import { drawLabel } from './sphereRenderer.js';
import { drawPhaseOverlay } from './moonPhaseRenderer.js';
import { drawClouds, drawLittleHouse } from './sceneryRenderer.js';

// Orchestrates the "view from Earth" sky scene.
export class SkySceneRenderer {
  constructor(starfield) {
    this._starfield = starfield;
  }

  drawEarthSkyView(ctx, x, y, w, h,
    sunElevation, moonAngleDeg, earthRotationDeg, illumination,
    phase, timeOfDay, hourOfDay) {

    const groundY = y + h * 0.75;

    // Sky gradient
    this._drawSkyGradient(ctx, x, y, w, h, groundY, sunElevation);

    // Ground
    this._drawGround(ctx, x, y, w, h, groundY, sunElevation);

    // Sun in the sky
    if (sunElevation > -0.1) {
      const sunSkyX = x + w * 0.5 + w * 0.35 * Math.cos(earthRotationDeg * Math.PI / 180.0);
      const sunSkyY = groundY - (groundY - y) * 0.8 * Math.max(0, sunElevation);

      if (sunElevation > 0) {
        const sunR = Math.min(w, h) * 0.06;

        // Sun glow
        ctx.save();
        ctx.shadowBlur = sunR * 2;
        ctx.shadowColor = `rgba(255,230,100,${0.16 * sunElevation})`;
        ctx.fillStyle = `rgba(255,230,100,${0.16 * sunElevation})`;
        ctx.beginPath();
        ctx.arc(sunSkyX, sunSkyY, sunR * 3, 0, Math.PI * 2);
        ctx.fill();
        ctx.restore();

        // Sun disk
        const sunGrad = ctx.createRadialGradient(sunSkyX, sunSkyY, 0, sunSkyX, sunSkyY, sunR);
        sunGrad.addColorStop(0, 'rgb(255,255,230)');
        sunGrad.addColorStop(1, 'rgb(255,200,50)');
        ctx.fillStyle = sunGrad;
        ctx.beginPath();
        ctx.arc(sunSkyX, sunSkyY, sunR, 0, Math.PI * 2);
        ctx.fill();
      }
    }

    // Moon in the sky
    const moonObserverAngle = (moonAngleDeg - earthRotationDeg) * Math.PI / 180.0;
    const moonElevation = Math.cos(moonObserverAngle);

    const minIlluminationToSee = sunElevation > 0.2 ? 0.25 : 0.03;
    const moonVisible = moonElevation > 0.05 && illumination > minIlluminationToSee;

    if (moonVisible) {
      const moonSkyX = x + w * 0.5 - w * 0.3 * Math.sin(moonObserverAngle);
      const moonSkyY = groundY - (groundY - y) * 0.7 * moonElevation;
      const moonR = Math.min(w, h) * 0.055;

      const moonAlpha = Math.min(1.0, moonElevation * 2) * Math.min(1.0, illumination * 3);

      this._drawMiniMoonPhase(ctx, moonSkyX, moonSkyY, moonR,
        moonAngleDeg, sunElevation, moonAlpha);
    }

    // Stars (only visible at night)
    if (sunElevation < 0.2) {
      const starAlpha = Math.max(0, Math.min(1, (0.2 - sunElevation) / 0.4));
      this._drawSkyStars(ctx, x, y, w, groundY, starAlpha);

      if (sunElevation < -0.2) {
        this._drawShootingStar(ctx, x, y, w, h, groundY, this._starfield.twinkleFrame);
      }
    }

    // Clouds
    drawClouds(ctx, x, y, w, h, groundY, sunElevation, this._starfield.twinkleFrame);

    // Little house
    drawLittleHouse(ctx, x, y, w, h, groundY, sunElevation);

    // Time label
    const hours = Math.floor(hourOfDay);
    const mins = Math.floor((hourOfDay - hours) * 60);
    const timeStr = `${String(hours).padStart(2, '0')}:${String(mins).padStart(2, '0')}`;
    drawLabel(ctx, timeStr, x + w * 0.5, y + 22, 16, [230, 230, 230]);

    // Day/Night label
    drawLabel(ctx, timeOfDay.name, x + w * 0.5, y + 44, 14, [200, 200, 220]);
  }

  _drawMiniMoonPhase(ctx, cx, cy, radius, moonAngleDeg, sunElevation, alpha) {
    // Moon glow
    const glowAlpha = (sunElevation < 0 ? 0.4 : 0.15) * alpha;
    ctx.save();
    ctx.shadowBlur = radius * 0.8;
    ctx.shadowColor = `rgba(200,200,220,${glowAlpha})`;
    ctx.fillStyle = `rgba(200,200,220,${glowAlpha})`;
    ctx.beginPath();
    ctx.arc(cx, cy, radius * 1.5, 0, Math.PI * 2);
    ctx.fill();
    ctx.restore();

    // Moon disk
    ctx.fillStyle = `rgba(220,220,215,${alpha})`;
    ctx.beginPath();
    ctx.arc(cx, cy, radius, 0, Math.PI * 2);
    ctx.fill();

    // Phase shadow
    const shadowAlpha = alpha * (sunElevation < 0 ? 230 : 200) / 255;
    const shadowColor = sunElevation < 0
      ? `rgba(5,5,25,${shadowAlpha})`
      : `rgba(100,140,200,${shadowAlpha})`;
    drawPhaseOverlay(ctx, cx, cy, radius, moonAngleDeg, shadowColor);
  }

  _drawSkyGradient(ctx, bx, by, bw, bh, groundY, sunElevation) {
    let topColor, bottomColor;

    if (sunElevation > 0.3) {
      topColor = [40, 100, 220];
      bottomColor = [135, 190, 250];
    } else if (sunElevation > 0.0) {
      const t = sunElevation / 0.3;
      topColor = blendColor([30, 30, 80], [40, 100, 220], t);
      bottomColor = blendColor([220, 100, 40], [135, 190, 250], t);
    } else if (sunElevation > -0.3) {
      const t = (sunElevation + 0.3) / 0.3;
      topColor = blendColor([8, 8, 35], [30, 30, 80], t);
      bottomColor = blendColor([20, 15, 40], [220, 100, 40], t);
    } else {
      topColor = [5, 5, 25];
      bottomColor = [10, 10, 35];
    }

    const grad = ctx.createLinearGradient(bx + bw / 2, by, bx + bw / 2, groundY);
    grad.addColorStop(0, `rgb(${Math.round(topColor[0])},${Math.round(topColor[1])},${Math.round(topColor[2])})`);
    grad.addColorStop(1, `rgb(${Math.round(bottomColor[0])},${Math.round(bottomColor[1])},${Math.round(bottomColor[2])})`);
    ctx.fillStyle = grad;
    ctx.fillRect(bx, by, bw, groundY - by);
  }

  _drawGround(ctx, bx, by, bw, bh, groundY, sunElevation) {
    let groundTop, groundBottom;

    if (sunElevation > 0.1) {
      groundTop = [40, 120, 40];
      groundBottom = [30, 80, 30];
    } else if (sunElevation > -0.1) {
      const t = (sunElevation + 0.1) / 0.2;
      groundTop = blendColor([15, 40, 15], [40, 120, 40], t);
      groundBottom = blendColor([10, 25, 10], [30, 80, 30], t);
    } else {
      groundTop = [15, 40, 15];
      groundBottom = [10, 25, 10];
    }

    const grad = ctx.createLinearGradient(bx + bw / 2, groundY, bx + bw / 2, by + bh);
    grad.addColorStop(0, `rgb(${Math.round(groundTop[0])},${Math.round(groundTop[1])},${Math.round(groundTop[2])})`);
    grad.addColorStop(1, `rgb(${Math.round(groundBottom[0])},${Math.round(groundBottom[1])},${Math.round(groundBottom[2])})`);
    ctx.fillStyle = grad;
    ctx.fillRect(bx, groundY, bw, by + bh - groundY);

    // Horizon line glow during sunrise/sunset
    if (sunElevation > -0.1 && sunElevation < 0.2) {
      ctx.save();
      ctx.shadowBlur = 8;
      const a = (60 * (1 - Math.abs(sunElevation) * 5)) / 255;
      ctx.shadowColor = `rgba(255,180,80,${a})`;
      ctx.strokeStyle = `rgba(255,180,80,${a})`;
      ctx.lineWidth = 1;
      ctx.beginPath();
      ctx.moveTo(bx, groundY);
      ctx.lineTo(bx + bw, groundY);
      ctx.stroke();
      ctx.restore();
    }
  }

  _drawSkyStars(ctx, bx, by, bw, groundY, alpha) {
    const stars = this._starfield.stars;
    if (!stars) return;

    for (let i = 0; i < 60; i++) {
      const sx = (stars[i].x % bw) + bx;
      const sy = (stars[i].y % (groundY - by)) + by;

      const twinkle = this._starfield.starBrightness[i] *
        (0.7 + 0.3 * Math.sin((this._starfield.twinkleFrame + i * 37) * 0.02));
      const a = twinkle * alpha;

      ctx.fillStyle = `rgba(255,255,240,${a})`;
      const size = this._starfield.starBrightness[i] > 0.8 ? 1.5 : 0.8;
      ctx.beginPath();
      ctx.arc(sx, sy, size, 0, Math.PI * 2);
      ctx.fill();
    }
  }

  _drawShootingStar(ctx, bx, by, bw, bh, groundY, frame) {
    const cycle = frame % 300;
    if (cycle > 30) return;

    const progress = cycle / 30;
    const startX = bx + bw * 0.7;
    const startY = by + 20;
    const endX = bx + bw * 0.3;
    const endY = groundY * 0.4;

    const curX = startX + (endX - startX) * progress;
    const curY = startY + (endY - startY) * progress;

    const tailLen = 30 * (1 - progress);
    const tailX = curX - ((endX - startX) / (endY - startY)) * -tailLen;
    const tailY = curY - tailLen;

    // Tail
    const tailGrad = ctx.createLinearGradient(tailX, tailY, curX, curY);
    tailGrad.addColorStop(0, 'rgba(255,255,220,0)');
    tailGrad.addColorStop(1, `rgba(255,255,220,${(200 * (1 - progress)) / 255})`);
    ctx.strokeStyle = tailGrad;
    ctx.lineWidth = 2;
    ctx.lineCap = 'round';
    ctx.beginPath();
    ctx.moveTo(tailX, tailY);
    ctx.lineTo(curX, curY);
    ctx.stroke();

    // Head glow
    ctx.save();
    ctx.shadowBlur = 3;
    ctx.shadowColor = `rgba(255,255,230,${(255 * (1 - progress)) / 255})`;
    ctx.fillStyle = `rgba(255,255,230,${(255 * (1 - progress)) / 255})`;
    ctx.beginPath();
    ctx.arc(curX, curY, 2, 0, Math.PI * 2);
    ctx.fill();
    ctx.restore();
  }
}
